### ISSUE TITLE
EC2 Fleet Resource

### ISSUE THREAD

**Issue by erks** [terraform-providers/terraform-provider-aws#4581]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

EC2 Fleet gives the ability to create fleets that are composed of a combination of EC2 On-Demand, Reserved, and Spot Instances with a single API call. 

https://aws.amazon.com/blogs/aws/ec2-fleet-manage-thousands-of-on-demand-and-spot-instances-with-one-request/

### New or Affected Resource(s)

* aws_ec2_fleet

### Potential Terraform Configuration

```hcl
resource "aws_launch_template" "foo" {
...
}

resource "aws_ec2_fleet" "foo" {
  request_type = "maintain"

  launch_template = {
    id = "${aws_launch_template.foo.id}"
    version = "$$Latest"
  }

  launch_template_overrides {
    instance_type = "m4.16xlarge"
    weighted_capacity = 64
  }

  launch_template_overrides {
    instance_type = "m5.24xlarge"
    weighted_capacity = 96
  }

  total_target_capacity = 2880
  on_demand_target_capacity = 960
  spot_target_capacity = 1920
  default_target_capacity_type = "spot"
}
```

### References

* https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet.html

**Comment by cread** [comment]
Anyone going to have a serious go at this in the near future?

**Comment by bflad** [comment]
I'll try to get an initial pull request up today. 😄

**Comment by bflad** [comment]
Initial pull request: #5960 -- implements most of the functionality, rest of which I can finish up on Monday hopefully.

**Comment by bflad** [comment]
The new `aws_ec2_fleet` resource has been merged and will release with version 1.39.0 of the AWS provider, likely later today. 👍
